portal/import_template.php

```diff
@@ -95,19 +98,29 @@
     // set proper permissions on the new file
     chmod(UPLOAD_DIR . $name, 0644);
     header("location: " . $_SERVER['HTTP_REFERER']);
+    die();
 }
 
 function validateFile($filename = '')
 {
-    $valid = false;
-    $filePath = $GLOBALS['OE_SITE_DIR'] . '/documents/onsite_portal_documents/templates/';
-    if (stripos($filename, $filePath) === false || !realpath($filename)) {
-        return false;
+    $knownPath = $GLOBALS['OE_SITE_DIR'] . '/documents/onsite_portal_documents/templates/'; // default path
+    $unknown = str_replace("\\", "/", realpath($filename)); // normalize requested path
+    $parts = pathinfo($unknown);
+    $unkParts = explode('/', $parts['dirname']);
+    $ptpid = $unkParts[count($unkParts) - 1]; // is this a patient or global template
+    $ptpid = ($ptpid == 'templates') ? '' : ($ptpid . '/'); // last part should be pid or template
+    $rebuiltPath = $knownPath . $ptpid . $parts['filename'] . '.tpl';
+    if (file_exists($rebuiltPath) === false || $parts['extension'] != 'tpl') {
+        redirect();
+    } elseif ($rebuiltPath != $unknown) { // these need to match to be valid request
```

Comment:
hi @sjpadgett ,
This looks great. The only thing I would add is if the above line should instead be:
```
} elseif (realpath($rebuiltPath) != realpath($filename)) { // these need to match to be valid request
```
This will then avoid issues in windows where C:\ could be added to the beginning of the realpath strings

Reply:
All paths are already fully qualified by U.I before being passed as these filepaths come from directory reads or built from Global file path variables. So drive is already part of path making realpath redundant. If you want to add, I suppose it can't hurt anything.

Reply:
ok, being super vigilant and added the following :)
(testing well)
https://github.com/openemr/openemr/pull/1765/commits/bcd2ca192bb0aa25e4b4f357f5b7ccb0fe263527